File: app/views/test/web_component/actions.js
import WebComponent from '../../../../lib/proscenium/ui/web_component'

class TestActions extends WebComponent {
  static componentName = 'pui-test-actions'
  static actions = {
    click: ['doClick'],
    change: ['doChange'],
    keydown: ['doKey']
  }

  #log(line) {
    const li = document.createElement('li')
    li.textContent = line
    this.querySelector('#action-log').appendChild(li)
  }

  doClick(event, target) {
    this.#log(`click:${target.id}:${event.target.id}`)
  }

  doChange(event, target) {
    this.#log(`change:${target.id}:${target.value}`)
  }

  doKey(event, target) {
    this.#log(`keydown:${target.id}:${event.key}`)
  }
}

WebComponent.register(TestActions)

File: app/views/test/web_component/actions.rb
# frozen_string_literal: true

module Views
  class Test::WebComponent::Actions < Application
    include Phlexible::Rails::AutoLayout

    sideload_assets js: { type: 'module' }

    register_element :pui_test_actions

    self.page_title = 'WebComponent — actions'

    def view_template
      h1 { 'Actions' }

      pui_test_actions(id: 'actions-host') do
        button(id: 'click-btn', 'on-click': 'doClick') { 'click' }
        input(id: 'change-input', type: 'text', 'on-change': 'doChange')
        input(id: 'key-input', type: 'text', 'on-keydown': 'doKey')
        div(id: 'click-target', 'on-click': 'doClick') do
          span(id: 'click-target-inner') { 'inner' }
        end
        ul(id: 'action-log')
      end
    end
  end
end
